tempesta_fw/client.c

```diff
@@ -28,121 +28,136 @@
-		HLIST_HEAD_INIT,
-	}
-};
-
-static struct kmem_cache *cli_cache;
-
-/* Total number of created clients. */
-static atomic64_t act_cli_n = ATOMIC64_INIT(0);
+static TDB *client_db;
 
 /**
  * Called when a client socket is closed.
  */
 void
 tfw_client_put(TfwClient *cli)
 {
+	TfwClientEntry *ent;
+	TdbRec *rec;
+
 	TFW_DBG2("put client %p, conn_users=%d\n",
 		 cli, atomic_read(&cli->conn_users));
 
 	if (!atomic_dec_and_test(&cli->conn_users))
 		return;
 
-	spin_lock(cli->hb_lock);
+	rec = (TdbRec *)((void *)cli - offsetof(TdbRec, data));
+	tdb_rec_get_lock(rec);
 
-	if (atomic_read(&cli->conn_users)) {
-		spin_unlock(cli->hb_lock);
-		return;
-	}
-	BUG_ON(!list_empty(&cli->conn_list));
-
-	hlist_del(&cli->hentry);
+	ent = (TfwClientEntry *)cli;
+	ent->expires = tfw_current_timestamp() + client_cfg.lifetime;
 
-	spin_unlock(cli->hb_lock);
+	tdb_rec_put(rec);
 
-	TFW_DBG("free client: cli=%p\n", cli);
-	kmem_cache_free(cli_cache, cli);
-	atomic64_dec(&act_cli_n);
+	TFW_DBG("put client: cli=%p\n", cli);
 	TFW_DEC_STAT_BH(clnt.online);
 }
-EXPORT_SYMBOL(tfw_client_put);
+
+static bool
+tfw_client_addr_eq(TdbRec *rec, void *data)
+{
+	TfwClientEntry *ent = (TfwClientEntry *)rec->data;
+	TfwClient *cli = &ent->cli;
+
+	return !memcmp_fast(&cli->addr.sin6_addr, data,
+			    sizeof(cli->addr.sin6_addr));
+}
 
 /**
- * Find a client corresponding to the @sk by IP address.
- * More advanced identification is possible based on User-Agent,
- * Cookie and other HTTP headers.
- * TODO (#488,#598,#1054?) actually clients can be relatively reliably
- * identified with TLS sessions (#1054) or HTTP sticky cookies.
+ * Find a client corresponding to @addr, @cli_addr (e.g. from X-Forwarded-For)
+ * and @user_agent.
  *
  * The returned TfwClient reference must be released via tfw_client_put()
  * when the @sk is closed.
- *
- * TODO #488 (#100): evict connections and/or clients and drop their accounting.
- * For now a client is freed immediately when the last of its connections is
- * closed, probably we should evict clients after some timeout to keep their
- * classifier statistic for following sessions...
+ * TODO #515 employ eviction strategy for the table
  */
 TfwClient *
-tfw_client_obtain(struct sock *sk, void (*init)(TfwClient *))
+tfw_client_obtain(TfwAddr addr, TfwAddr *cli_addr, TfwStr *user_agent,
+		  void (*init)(TfwClient *))
 {
+	TfwClientEntry *ent;
 	TfwClient *cli;
-	CliHashBucket *hb;
-	struct hlist_node *tmp;
 	unsigned long key;
-	TfwAddr addr;
+	size_t len;
+	TdbRec *rec;
+	bool is_new;
+	int conn_users;
+	time_t curr_time = tfw_current_timestamp();
 
-	ss_getpeername(sk, &addr);
 	key = hash_calc((const char *)&addr.sin6_addr,
 			sizeof(addr.sin6_addr));
 
-	hb = &cli_hash[hash_min(key, CLI_HASH_BITS)];
-
-	spin_lock(&hb->lock);
+	if (cli_addr)
+		key ^= hash_calc((const char *)&cli_addr->sin6_addr,
+				 sizeof(cli_addr->sin6_addr));
+	if (user_agent) {
+		char buf[256];
+		size_t len = tfw_str_to_cstr(user_agent, buf, sizeof(buf));
```

Comment:
You spin over TfwStr chunks twice: first, copying string to buffer, second time - during hash calculation. This is not very good. Copy operation should be avoided. See https://github.com/tempesta-tech/tempesta/blob/5adf46dd9b94c9fa67bdd0233e3b836bc4931ad9/tempesta_fw/hash.c#L33 I've wrote about this function.

Reply:
User-Agent can be very large. It will slowly work

Reply:
The comment isn't addressed. It makes sense to limit how many bytes we use for hashing, but it'd appreciated if you extend `tfw_hash_str()` with a length of data to process from `str`. The copying isn't nice at all.